core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -135,190 +142,498 @@ private SharePartitionKey sharePartitionKey(String groupId, TopicIdPartition top
+                    if (session.epoch != expectedEpoch) {
+                        log.info("Subsequent share session {} expected epoch {}, but got {}. " +
+                                "Possible duplicate request.", session.key, expectedEpoch, session.epoch);
+                        return new ShareFetchResponse(ShareFetchResponse.toMessage(Errors.INVALID_SHARE_SESSION_EPOCH,
+                                throttleTimeMs, Collections.emptyIterator(), Collections.emptyList()));
+                    } else
+                        return new ShareFetchResponse(ShareFetchResponse.toMessage(Errors.NONE, throttleTimeMs,
+                                Collections.emptyIterator(), Collections.emptyList()));
+                }
+            }
         }
-    }
 
-    /**
-     * The share fetch context for a share fetch request that had a session error.
-     */
-    // TODO: Implement ShareSessionErrorContext when you have share sessions available
-    public static class ShareSessionErrorContext extends ShareFetchContext {
+        // Iterator that goes over the given partition map and selects partitions that need to be included in the response.
+        // If updateShareContextAndRemoveUnselected is set to true, the share context will be updated for the selected
+        // partitions and also remove unselected ones as they are encountered.
+        private class PartitionIterator implements Iterator<Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData>> {
+            private final Iterator<Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData>> iterator;
+            private final boolean updateShareContextAndRemoveUnselected;
+            private Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData> nextElement;
+
+
+            public PartitionIterator(Iterator<Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData>> iterator, boolean updateShareContextAndRemoveUnselected) {
+                this.iterator = iterator;
+                this.updateShareContextAndRemoveUnselected = updateShareContextAndRemoveUnselected;
+            }
+
+            @Override
+            public boolean hasNext() {
+                while ((nextElement == null) && iterator.hasNext()) {
+                    Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData> element = iterator.next();
+                    TopicIdPartition topicPart = element.getKey();
+                    ShareFetchResponseData.PartitionData respData = element.getValue();
+                    CachedSharePartition cachedPart = session.partitionMap.find(new CachedSharePartition(topicPart));
+                    boolean mustRespond = cachedPart.maybeUpdateResponseData(respData);
+                    if (mustRespond) {
+                        nextElement = element;
+                        if (updateShareContextAndRemoveUnselected && ShareFetchResponse.recordsSize(respData) > 0) {
```

Comment:
`session.partitionMap` is of type `ImplicitLinkedHashCollection<>` which tracks the order of insertion of elements. Since, we are updating an element in this case, we need to perform a `remove` and then a `mustAdd` to maintain the correct order